### PR TITLE
feat: smart currency choosing

### LINKED ISSUE #702: Smart currency choosing
## Description 

The idea is that we could choose the default currency based on the user's balance. We would check all balances and set the payment currency to the biggest USD-worth balance. 

I think we always to prioritize ETH for one time payment but in cases of subscription it could be very very useful as people tend to keep more STRK than ETH on Starknet for example.

I would propose to only do it on `renewal.tsx` + `subscription.tsx` & `register.tsx`

### PR DESCRIPTION
**Pull request by irisdv**
Close #702

### CODE REVIEW

File: components/discount/renewalDiscount.tsx [hunk 1]
```diff
@@ -103,29 +101,24 @@ const RenewalDiscount: FunctionComponent<RenewalDiscountProps> = ({
...
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({
+        meta_hash: metadataHash,
+        email: email,
+        tax_state: isSwissResident ? "switzerland" : "none",
+        salt: salt,
+      }),
+    })
+      .then((res) => res.json())
+      .catch((err) => console.log("Error on sending metadata:", err));
```
Comment (fricoben) [inline components/discount/renewalDiscount.tsx]: Why do you pass this as something with only one hash, I don't understand the link with smart currency choosing ?

File: components/domains/steps/checkoutCard.tsx [hunk 2]
```diff
@@ -174,6 +175,16 @@ const CheckoutCard: FunctionComponent<CheckoutCardProps> = ({
     discount,
   ]);
 
+  // we choose the currency based on the user balances
+  useEffect(() => {
+    if (tokenBalances && !hasChoseCurrency) {
```
Comment (fricoben) [inline components/domains/steps/checkoutCard.tsx]: `hasChosenCurrency`*

File: utils/altcoinService.ts [hunk 3]
```diff
@@ -115,3 +115,45 @@ export const getAutoRenewAllowance = (
...
   return allowance;
 };
+
+type AvnuQuote = {
+  address: string;
+  currentPrice: number;
+  decimals: number;
+};
+
+export const smartCurrencyChoosing = async (
```
Comment (fricoben) [inline utils/altcoinService.ts:133]: I think this function is a bit difficult to understand as one block, I asked chatGPT 4 to refactor it, here is a proposition he made: 

```ts
enum CurrencyType {
  ETH = 'ETH',
  STRK = 'STRK'
}

interface TokenBalances {
  ETH: string;
  STRK: string;
}

interface AvnuQuote {
  address: string;
  currentPrice: number;
  decimals: number;
}

// Helper to convert hexadecimal to decimal
const hexToDecimal = (hex: string): number => parseInt(hex, 16);

// API call to fetch quote data
async function fetchAvnuQuoteData(): Promise<AvnuQuote[]> {
  const response = await fetch(`${process.env.NEXT_PUBLIC_AVNU_API}/tokens/short?in=0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7`);
  return response.json();
}

// Determine the currency type based on token balances and STRK quote
async function determineCurrencyType(tokenBalances: TokenBalances): Promise<CurrencyType> {
  // Early returns based on token presence
  if (tokenBalances.ETH && !tokenBalances.STRK) return CurrencyType.ETH;
  if (!tokenBalances.ETH) return CurrencyType.STRK;

  // Fetch data and find relevant quote
  const avnuQuotes = await fetchAvnuQuoteData();
  const contractAddress = ERC20Contract[CurrencyType.STRK];
  const quoteData = avnuQuotes.find(quote => hexToDecimal(quote.address) === hexToDecimal(contractAddress));

  // Fallback to ETH if no STRK quote data is found
  if (!quoteData) return CurrencyType.ETH;

  // Convert STRK balance to ETH equivalent
  const strkBalance = BigInt(tokenBalances.STRK);
  const strkQuote = BigInt(Math.round(quoteData.currentPrice * Math.pow(10, quoteData.decimals)));
  const strkConvertedBalance = (strkBalance * strkQuote) / BigInt(Math.pow(10, quoteData.decimals));

  // Compare converted STRK balance to ETH balance
  return BigInt(tokenBalances.ETH) * BigInt(3) < strkConvertedBalance ? CurrencyType.STRK : CurrencyType.ETH;
}

// Example usage:
const tokenBalances: TokenBalances = { ETH: '100', STRK: '200' };
determineCurrencyType(tokenBalances).then(currency => console.log(currency));

```

I think it's way cleaner :), and it'll permit you to have easier tests with better coverage! Don't forget to test :)

File: utils/discounts/freeRenewal.ts [hunk 4]
```diff
@@ -10,7 +10,7 @@ type FreeRenewalDiscount = {
 export const freeRenewalDiscount: FreeRenewalDiscount = {
   name: "The Free Renewal",
   image: "/freeRenewal/freeRenewal.webp",
-  expiry: 1702598399 * 1000, // timestamp in ms
+  expiry: 1716634160 * 1000, // timestamp in ms
```
Comment (fricoben) [inline utils/discounts/freeRenewal.ts]: to delete cause we'll add all infos later

**Review by fricoben — CHANGES_REQUESTED** [review]
Some changes on the code and questions

File: components/discount/renewalDiscount.tsx (hunk 1 above)
Comment (irisdv) [reply components/discount/renewalDiscount.tsx]: It's linked to the `freeRenewal` page that uses the `renewalDiscount` component. 
It's an update we made for it to work with the subscriptions, now a user has only one `metadatahash`. And actually we should only store it if `needMetadata = true` so I will update it.

File: utils/altcoinService.ts [hunk 5]
```diff
@@ -115,3 +115,48 @@ export const getAutoRenewAllowance = (
...
+
+type AvnuQuote = {
+  address: string;
+  currentPrice: number;
+  decimals: number;
+};
+
+export async function fetchAvnuQuoteData(): Promise<AvnuQuote[]> {
+  const response = await fetch(
+    `${process.env.NEXT_PUBLIC_AVNU_API}/tokens/short?in=0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7`
```
Comment (fricoben) [inline utils/altcoinService.ts:127]: Might want to use `TOKEN_ADDRESS_MAP` here ?

**Comment by fricoben** [comment]
## To Test 

- [x] No STRK, give ETH as base currency
- [x] No ETH, no STRK, gives ETH as a base currency
- [x] X ETH, 4X STRK, gives STRK
- [x] X ETH, 2X STRK, gives ETH
- [x] Not enough STRK gives ETH
- [x] Not enough ETH gives STRK

**Review by fricoben — APPROVED** [review]
lgtm

**Pull request merged**